### PR TITLE
Add Lie Group Support to PyTorch Optimizers

### PR DESCRIPTION
**Pull request by fantaosha**
## Motivation and Context

Apply `torch.optim` to run optimization and train neural networks on Lie groups.

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have completed my CLA (see **CONTRIBUTING**)
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: examples/se2_inverse.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
+# This example will use torch to solve min_x1 || x1^-1 - x2 ||^2,
```
Comment (mhmukadam) [inline examples/se2_inverse.py:1]: Add license add the top and remember to update the readme in examples folder as necessary.

File: theseus/geometry/utils/utils.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+import torch
+
+
+class LieGroupTensor(torch.Tensor):
```
Comment (mhmukadam) [inline theseus/geometry/utils/utils.py:9]: I am wondering if we need this class to be in a separate utils folder (i.e. do we anticipate more stuff) or we can just have all this in theseus/geometry.

File: examples/se2_inverse.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+# where x1 and x2 are SE2 objects
+
+import torch
+import theseus as th
+from theseus import LieGroupTensor
+from theseus.geometry.lie_group import LieGroup
+
+from theseus.geometry.tests.test_se2 import create_random_se2
```
Comment (mhmukadam) [inline examples/se2_inverse.py:9]: We could move more generally useful stuff like create_random_se2 from tests to theseus/utils.

File: examples/se2_inverse.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+def run(x1: LieGroup, x2: LieGroup, num_iters=10, use_lie_tangent=True):
+    if use_lie_tangent:
+        # update x1 with the Lie group tangent gradient
+        x1.data.__class__ = LieGroupTensor
+        x1.data.__dict__["group_cls"] = type(x1)
+    else:
+        # update x1 with the Euclidean gradient
+        x1.data.__class__ = torch.Tensor
+        x1.data.__dict__.pop("group_cls", None)
```
Comment (mhmukadam) [inline examples/se2_inverse.py:25]: I think if we can find a way to add this logic to the LieGroup class itself and use a variable (like dtype or device) to specific which grad mode to use then the overall api would be a lot more user friendly.

**Review by mhmukadam — COMMENTED** [review]
Nice work! We should also add an example/test showing usage with a TheseusLayer objective defined on Lie group variables and an inner nonlinear optimizer and an outer pytorch optimizer.

File: examples/se2_inverse.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        loss = cf.error().norm()
+        if i % 100 == 0:
+            print(
+                "iter {:0>4d}: loss is {:.10f}, cos(theta)^2 + sin(theta)^2 is {:.10f}".format(
+                    i, loss.item(), x1[0, 2:].norm().item() ** 2
+                )
+            )
+        loss.backward()
+
+        with th.set_lie_tangent_enabled(use_lie_tangent):
```
Comment (fantaosha) [inline examples/se2_inverse.py:43]: The users need this line to update `LieGroup` on its tangent space not the Euclidean space.

File: examples/se2_inverse.py [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+from theseus.geometry.tests.test_se2 import create_random_se2
+
+rng = torch.Generator()
+rng.manual_seed(0)
+x1 = create_random_se2(1, rng)
+x2 = create_random_se2(1, rng)
+
+
+def run(x1: LieGroup, x2: LieGroup, num_iters=10, use_lie_tangent=True):
+    x1.data = LieGroupTensor(x1)
```
Comment (fantaosha) [inline examples/se2_inverse.py:24]: We don't need this line if `theseus.LieGroup.data` is changed from `torch.Tensor` to `theseus.LieGroupTensor`.

File: examples/se2_inverse.py (hunk 4 above)
Comment (fantaosha) [reply examples/se2_inverse.py]: > I think if we can find a way to add this logic to the LieGroup class itself and use a variable (like dtype or device) to specific which grad mode to use then the overall api would be a lot more user friendly.

I'm OK for either.

File: examples/se2_inverse.py (hunk 1 above)
Comment (fantaosha) [reply examples/se2_inverse.py:7]: Updated.

File: theseus/geometry/utils/utils.py (hunk 2 above)
Comment (fantaosha) [reply theseus/geometry/utils/utils.py:57]: I have no preference and I'm OK for this.

File: theseus/geometry/utils/utils.py [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+# LICENSE file in the root directory of this source tree.
+
+import threading
+from typing import Any
+
+import torch
+
+
+class LieGroupContext(object):
+    contexts = threading.local()
```
Comment (fantaosha) [inline theseus/geometry/utils/utils.py:13]: Someone said that `threading.local()` is safe for multi-thread processing. Are there any other concerns with this? @luisenp @mhmukadam

File: examples/se2_inverse.py (hunk 3 above)
Comment (luisenp) [reply examples/se2_inverse.py:15]: Created #78 for this.

File: theseus/geometry/utils/utils.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+        self.prev = super().get_context()
+        LieGroupContext.set_context(False)
+        return self
+
+    def __exit__(self, typ, value, traceback):
+        LieGroupContext.set_context(self.prev)
+
+
+class LieGroupTensor(torch.Tensor):
+    from torch._C import _disabled_torch_function_impl
```
Comment (luisenp) [inline theseus/geometry/utils/utils.py:58]: Is this related to [this](https://github.com/pytorch/pytorch/issues/46159#issuecomment-707817037) issue? Can you explain a bit why this part is necessary?

File: theseus/geometry/utils/utils.py (hunk 2 above)
Comment (luisenp) [reply theseus/geometry/utils/utils.py:57]: I think we can keep these in the same folder, as @mhmukadam suggests. On the other hand, we want to discourage users from using these classes directly, so we should rename as `_LieGroupTensor` and `_LieGroupContext`.

File: theseus/geometry/utils/utils.py [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+
+    def __new__(cls, group):
+        return torch.Tensor._make_subclass(cls, group.data)
+
+    def __init__(self, group):
+        self.group_cls = type(group)
+
+    def add_(self, update, alpha=1):
+        if LieGroupContext.get_context():
+            group = self.group_cls(data=self.data)
```
Comment (luisenp) [inline theseus/geometry/utils/utils.py:70]: One concern here and for the functions below is that this hard codes the assumption that a call to `add_` wants to add the gradient, which might not be the case for all of the optimization algorithms in Python. 

Is there a way to change this implementation so that within the context `lie_tensor.grad` value is replaced by its projection to tangent space, and `add_` is just a call to `retract`?

File: theseus/geometry/utils/utils.py (hunk 7 above)
Comment (luisenp) [reply theseus/geometry/utils/utils.py:13]: Nothing comes to mind right now, but we should keep an eye open and tests this thoroughly.

File: theseus/geometry/utils/utils.py [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+        if not hasattr(cls.contexts, "use_lie_tangent"):
+            cls.contexts.use_lie_tangent = False
+        return cls.contexts.use_lie_tangent
+
+    @classmethod
+    def set_context(cls, use_lie_tangent: bool):
+        cls.contexts.use_lie_tangent = use_lie_tangent
+
+
+class set_lie_tangent_enabled(object):
```
Comment (luisenp) [inline theseus/geometry/utils/utils.py:26]: Could we rename these as `set_tangent_enabled`, `enable_tangent`, `no_tangent`? (dropping the "lie"). Is that unambiguous enough? Since these context will be used a lot, I'd prefer shorter names.

cc: @mhmukadam

File: examples/se2_inverse.py [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+    x1.data.requires_grad = True
+
+    optim = torch.optim.Adam([x1.data], lr=1e-1)
```
Comment (luisenp) [inline examples/se2_inverse.py:27]: Once we converge on a solution we are satisfied with, we should test this with all of the optimizer under `torch.optim`.

File: examples/se2_inverse.py (hunk 6 above)
Comment (luisenp) [reply examples/se2_inverse.py:24]: I think this should be fine, but let's wait until we finalize the implementation of `LieGroupTensor`.

**Review by luisenp — CHANGES_REQUESTED** [review]
This is a clever workaround! I left some important questions, and we need to test this carefully to avoid unwanted behavior, but this is promising.

File: theseus/geometry/utils/utils.py (hunk 2 above)
Comment (fantaosha) [reply theseus/geometry/utils/utils.py:57]: Agreed. Will make this change.

File: theseus/geometry/utils/utils.py (hunk 9 above)
Comment (fantaosha) [reply theseus/geometry/utils/utils.py:70]: That's what I previously tried. However, we might might implement `torch.optim` algorithms in that case. This is something similar to the notion of `virtual function` and `retract()` will be called only if`LieGroupContext.get_context()` is `True`.

File: examples/se2_inverse.py (hunk 11 above)
Comment (fantaosha) [reply examples/se2_inverse.py:27]: I created a new issue #79.

File: theseus/geometry/utils/utils.py (hunk 10 above)
Comment (fantaosha) [reply theseus/geometry/utils/utils.py:26]: My concern is that the Euclidean gradient is mathematically on the Euclidean tangent space, while the Lie tangent gradient is on the Lie group tangent space. Both of them are the tangent spaces.

File: theseus/geometry/utils/utils.py (hunk 9 above)
Comment (fantaosha) [reply theseus/geometry/utils/utils.py:70]: BTW, note that `LieGroupContext.get_context()` is `False` by default and the users need something like `with theseus.lie_tangent_enabled` to trigger `retract`.

File: theseus/geometry/utils/utils.py (hunk 10 above)
Comment (luisenp) [reply theseus/geometry/utils/utils.py:26]: Ah, OK, that makes sense. Then it's probably safest to keep the "lie" prefix.

File: theseus/geometry/utils/utils.py (hunk 9 above)
Comment (luisenp) [reply theseus/geometry/utils/utils.py:70]: Yes, I'm talking about the case where the `LieGroupContext.get_context()` is `True`. Even in that scenario, we don't know if some torch optimizer step will try to do an in-place addition whose argument is not the gradient, but some other vector (e.g. ,a different descent direction), and in this case doing the projection would be wrong, no? Or is this projection correct for any vector in eucildean space, not just the exact gradient?

File: examples/se2_inverse.py (hunk 11 above)
Comment (luisenp) [reply examples/se2_inverse.py:27]: I'd suggest doing these tests as part of this PR (but once we agree on the final solution).

Comment (fantaosha) [reply examples/se2_inverse.py:27]: Agreed.

File: theseus/geometry/utils/utils.py (hunk 9 above)
Comment (fantaosha) [reply theseus/geometry/utils/utils.py:70]: See #79

File: theseus/geometry/utils/utils.py (hunk 2 above)
Comment (fantaosha) [reply theseus/geometry/utils/utils.py]: See #93 and I will create a new PR for it.

File: theseus/geometry/utils.py [hunk 12]
```diff
@@ -0,0 +1,92 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
```
Comment (fantaosha) [inline theseus/geometry/utils.py:1]: I move `theseus/geometry/utils/utils.py` to `theseus/geometry/utils.py`

File: theseus/geometry/utils/utils.py (hunk 7 above)
Comment (mhmukadam) [reply theseus/geometry/utils/utils.py]: Captured in #92.
